File: frontend/src/api/analysis.ts
import axios from "axios"

const api = axios.create({
  baseURL: import.meta.env.VITE_API_URL ?? (import.meta.env.PROD ? "" : "http://localhost:8000"),
})

export interface Transaction {
  id: number | string
  amount: number
  merchant: string
  date: string
}

export interface Flag {
  transaction_id: string | number
  reason: string
  severity: "Low" | "Medium" | "High" | "Critical"
}

export interface AnalysisResult {
  id: string
  risk_score: number
  risk_level: "Low" | "Medium" | "High" | "Critical"
  summary: string
  flags: Flag[]
  recommendations: string[]
}

export interface Analysis {
  id: string
  created_at: string
  account_name: string
  risk_score: number
  risk_level: "Low" | "Medium" | "High" | "Critical"
  summary: string
  total_transactions: number
  flagged_transactions: number
}

export interface AnalysisDetail extends Analysis {
  flags: Flag[]
  recommendations: string[]
  transactions: Transaction[]
}

export interface DashboardSummary {
  total_analyses: number
  total_transactions: number
  flagged_transactions: number
  average_risk_score: number
  risk_level: "Low" | "Medium" | "High" | "Critical"
}

export const analyseTransactions = async (
  accountName: string,
  transactions: Transaction[]
): Promise<AnalysisResult> => {
  const { data } = await api.post("/api/analyze", { account_name: accountName, transactions })
  return data
}

export const getAccounts = async (): Promise<string[]> => {
  const { data } = await api.get("/api/accounts")
  return data
}

export const getAnalyses = async (account?: string): Promise<Analysis[]> => {
  const { data } = await api.get("/api/analyses", { params: account ? { account } : {} })
  return data
}

export const getAnalysisDetail = async (id: string): Promise<AnalysisDetail> => {
  const { data } = await api.get(`/api/analyses/${id}`)
  return data
}

export const getDashboardSummary = async (account?: string): Promise<DashboardSummary> => {
  const { data } = await api.get("/api/dashboard", { params: account ? { account } : {} })
  return data
}

export const askAboutAnalysis = async (analysisId: string, question: string): Promise<string> => {
  const { data } = await api.post(`/api/analyses/${analysisId}/ask`, { text: question })
  return data.answer
}

File: backend/app/routes/analysis.py
import uuid
from datetime import datetime, timezone

from fastapi import APIRouter, HTTPException, Depends, Query
from sqlalchemy.orm import Session

from app.schemas import AnalyseRequest
from app.services.analysis_service import (
    risk_service,
    AnalysisServiceError,
    MalformedResponseError,
)
from app.models.db_models import AnalysisDB
from app.database import get_db

router = APIRouter(prefix="/api", tags=["analysis"])


def utcnow() -> datetime:
    return datetime.now(timezone.utc)


@router.post("/analyze")
def analyse(payload: AnalyseRequest, db: Session = Depends(get_db)):
    try:
        transactions = [t.model_dump() for t in payload.transactions]
        result = risk_service.analyse_transactions(transactions)

        analysis_id = str(uuid.uuid4())
        analysis = AnalysisDB(
            id=analysis_id,
            created_at=utcnow(),
            account_name=payload.account_name,
            transactions=transactions,
            risk_score=result["risk_score"],
            risk_level=result["risk_level"],
            summary=result["summary"],
            flags=result["flags"],
            recommendations=result["recommendations"],
            total_transactions=len(transactions),
            flagged_transactions=len(result["flags"]),
        )
        db.add(analysis)
        db.commit()

        return {**result, "id": analysis_id}
    except MalformedResponseError as e:
        raise HTTPException(status_code=502, detail=str(e))
    except AnalysisServiceError as e:
        raise HTTPException(status_code=502, detail=str(e))
    except Exception:
        raise HTTPException(status_code=500, detail="Analysis failed")


@router.get("/accounts")
def get_accounts(db: Session = Depends(get_db)):
    rows = db.query(AnalysisDB.account_name).distinct().all()
    return sorted([r[0] for r in rows])


@router.get("/dashboard")
def get_dashboard(account: str | None = Query(None), db: Session = Depends(get_db)):
    query = db.query(AnalysisDB)
    if account:
        query = query.filter(AnalysisDB.account_name == account)
    analyses = query.order_by(AnalysisDB.created_at.desc()).all()

    if not analyses:
        return {
            "total_analyses": 0,
            "total_transactions": 0,
            "flagged_transactions": 0,
            "average_risk_score": 0,
            "risk_level": "Low",
        }

    latest = analyses[0]
    return {
        "total_analyses": len(analyses),
        "total_transactions": sum(a.total_transactions for a in analyses),
        "flagged_transactions": sum(a.flagged_transactions for a in analyses),
        "average_risk_score": round(sum(a.risk_score for a in analyses) / len(analyses)),
        "risk_level": latest.risk_level,
    }


@router.get("/analyses")
def get_analyses(account: str | None = Query(None), db: Session = Depends(get_db)):
    query = db.query(AnalysisDB)
    if account:
        query = query.filter(AnalysisDB.account_name == account)
    analyses = query.order_by(AnalysisDB.created_at.desc()).all()
    return [
        {
            "id": a.id,
            "created_at": a.created_at.isoformat(),
            "account_name": a.account_name,
            "risk_score": a.risk_score,
            "risk_level": a.risk_level,
            "total_transactions": a.total_transactions,
            "flagged_transactions": a.flagged_transactions,
            "summary": a.summary,
        }
        for a in analyses
    ]


@router.get("/analyses/{analysis_id}")
def get_analysis_detail(analysis_id: str, db: Session = Depends(get_db)):
    analysis = db.query(AnalysisDB).filter(AnalysisDB.id == analysis_id).first()
    if not analysis:
        raise HTTPException(status_code=404, detail="Analysis not found")
    return {
        "id": analysis.id,
        "created_at": analysis.created_at.isoformat(),
        "account_name": analysis.account_name,
        "risk_score": analysis.risk_score,
        "risk_level": analysis.risk_level,
        "summary": analysis.summary,
        "flags": analysis.flags,
        "recommendations": analysis.recommendations,
        "transactions": analysis.transactions,
        "total_transactions": analysis.total_transactions,
        "flagged_transactions": analysis.flagged_transactions,
    }


@router.get("/health")
def health():
    return {"status": "ok", "model": "llama-3.1-8b-instant"}
